### PR TITLE
process trait/impl items directly from the visitor callback

### PR DESCRIPTION
**Pull request by nikomatsakis**
The current setup processes impl/trait items while visiting
the impl/trait. This means we basically have this setup:

    <Lots> -> TypeckItemBody(Impl) -> Tables(ImplItem{0,1,2,3})

But this was largely an artifact of the older code. By moving the
processing of items into method dedicated for their use, we produce this
setup:

    <Little> -> TypeckItemBody(ImplItem0) -> Tables(ImplItem0)
    ...
    <Little> -> TypeckItemBody(ImplItem3) -> Tables(ImplItem3)

r? @michaelwoerister 

Also, we might consider removing the `TypeckItemBody` node altogether and just using `Tables` as the task. `Tables` is its primary output, I imagine? That would reduce size of dep-graph somewhat.

cc @eddyb -- perhaps this pattern applies elsewhere?

### CODE REVIEW

**Comment by eddyb** [comment]
LGTM. I agree with `Tables`, that was my intent.

**Comment by nikomatsakis** [comment]
Not as an argument against eliminating `TypeckItemBody` exactly, but it's worth noting that type-checking the item body produces *primarily* a `Tables` but also a `TypeckItemSignature` for the various closures in the body.

**Comment by eddyb** [comment]
The whole closure situation is a bit troublesome. It'd be better to actually compute (copy, really) that information from the surrounding function's `Tables` (and I believe I should've just done that instead of what it is now).

File: src/librustc/dep_graph/README.md [hunk 1]
```diff
@@ -440,4 +440,6 @@ To achieve this, the HIR map will detect if the def-id originates in
 an inlined node and add a dependency to a suitable `MetaData` node
 instead. If you are reading a HIR node and are not sure if it may be
 inlined or not, you can use `tcx.map.read(node_id)` and it will detect
-whether the node is inlined or not and do the right thing.
+whether the node is inlined or not and do the right thing.  You can
+also use `tcx.map.is_inlined_def_id()` and
+`tcx.map.is_inlined_node_id()` to test.
```
Comment (eddyb) [inline src/librustc/dep_graph/README.md]: None of this happens anymore, inlined bodies don't get their own local `NodeId`s or `DefId`s.

**Review by eddyb — CHANGES_REQUESTED** [review]
LGTM except one doc nit.

File: src/librustc/dep_graph/README.md (hunk 1 above)
Comment (michaelwoerister) [inline src/librustc/dep_graph/README.md]: Agreed.

**Comment by michaelwoerister** [comment]
Looks good to me! One nit: "Tables" is a rather generic name for the dep-node that says little about what it represents. If you can think of a more descriptive name, now would be the time to change it, I guess.
Not a priority though. r=me with @eddyb's comment addressed.

**Comment by eddyb** [comment]
`TypedHirTables` is maybe the most descriptive sequence of 3 words I can come up with.

**Review by eddyb — APPROVED** [review]

**Pull request merged**